### PR TITLE
[Sync II] Add ttl parsing tools

### PR DESCRIPTION
**Pull request by CodyCBakerPhD**
## Motivation
Extension of #245 that includes tools for parsing the TTL signal to identify the rising and falling frames. Precursor to sync methods of #237.

## How to test the behavior?
Added tests using the synthetic TTLs. Very easy to add new tests for different cases with this approach.

## Checklist

- [X] Have you thoroughly read our [Developer Guide](https://neuroconv.readthedocs.io/en/main/developer_guide.html)?
- [X] Have you ensured the PR description clearly describes the problem and solutions?
- [X] Have you checked to ensure that there aren't other open or previously closed [Pull Requests](https://github.com/catalystneuro/neuroconv/pulls) for the same change?
- [X] If this PR fixes an issue, is the first line of the PR description `fix #XXX` where `XXX` is the issue number?
- [X] Did you update the [CHANGLOG.md](https://github.com/catalystneuro/neuroconv/blob/main/CHANGELOG.md) file on your branch to describe your changes?

### CODE REVIEW

**Comment by CodyCBakerPhD** [comment]
From https://github.com/catalystneuro/neuroconv/pull/237#discussion_r1044871549,

I'd say given the amount of common pre-processing involved in calculating both, as well as the fact that future extensions of this work to harnessing TTL durations and values for decoding digital words, there's no harm at all in calculating both and only using one when needed. The extra `np.where` step, even on a long series, is not such a slow down that it would be cumbersome relative to the total time of parsing the rest of the TTL.

File: tests/test_minimal/test_tools/test_signal_processing.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        Default generation tests start in 'off' condition.
+
+        What should 'correct' behavior here be?
+
+        We searh for rising frames to count as a 'start' signal, but if the signal starts in an 'on' condition,
+        how to interpret?
+        """
+        ttl_signal = generate_mock_ttl_signal(signal_duration=5.0, ttl_times=[2.5], ttl_duration=5.0)
+
+        rising_frames, falling_frames = parse_rising_and_falling_frames_from_ttl(trace=ttl_signal)
```
Comment (bendichter) [inline tests/test_minimal/test_tools/test_signal_processing.py:51]: can you test the falling_frames is empty?

**Review by bendichter — CHANGES_REQUESTED** [review]

File: tests/test_minimal/test_tools/test_signal_processing.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+from hdmf.testing import TestCase
+from numpy.testing import assert_array_equal
+
+from neuroconv.tools.signal_processing import parse_rising_and_falling_frames_from_ttl
+from neuroconv.tools.testing import generate_mock_ttl_signal
+
+
+class TestParsingRisingAndFallingTimesFromTTL(TestCase):
+    def test_default(self):
+        ttl_signal = generate_mock_ttl_signal()
```
[the remark this answers is not in the document]
Comment (CodyCBakerPhD) [reply tests/test_minimal/test_tools/test_signal_processing.py:11]: Added a new test for this.

One with explicitly set values corresponding to the current default values.

The other that actually tests the default call (no arguments passed) to the function matches our expectations.

File: src/neuroconv/tools/signal_processing.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        The default is 2, signifying a clear on/off signal from a baseline value to a target value.
+
+    Returns
+    -------
+    rising_frames: numpy.ndarray
+        The frame indices of rising events.
+    falling_frames: numpy.ndarray
+        The frame indices of falling events.
+    """
+    eps = 1 if np.issubdtype(trace.dtype, np.integer) else np.finfo(float).eps
```
[the remark this answers is not in the document]
Comment (CodyCBakerPhD) [reply src/neuroconv/tools/signal_processing.py]: Adjusted to use the midpoint by default, and also allowing `bins` to be set explicitly if the user wants to assume manual control or use a custom threshold

File: src/neuroconv/tools/signal_processing.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
+"""Author: Cody Baker."""
+from typing import Tuple
+
+import numpy as np
+
+
+def parse_rising_and_falling_frames_from_ttl(trace: np.ndarray, num_bins: int = 2) -> Tuple[np.ndarray, np.ndarray]:
```
[the remark this answers is not in the document]
Comment (CodyCBakerPhD) [reply src/neuroconv/tools/signal_processing.py]: Is now two functions

[the remark this answers is not in the document]
Comment (CodyCBakerPhD) [reply src/neuroconv/tools/signal_processing.py]: Adjusted to simple 'get'

File: CHANGELOG.md [hunk 5]
```diff
@@ -40,6 +40,7 @@
 * Simplified the writing of `channel_conversion` during `add_electrical_series` if the vector of gains is uniform; in this case, they are now combined into the scalar `conversion` value. [PR #218](https://github.com/catalystneuro/neuroconv/pull/218)
 * Implement timestamp extraction from videos for the SLEAPInterface [PR #238](https://github.com/catalystneuro/neuroconv/pull/238)
 * Prevented writing of default values for optional columns on the `ElectrodeTable`. [PR #219](https://github.com/catalystneuro/neuroconv/pull/219)
+* Added a tool for determining rising and falling frames from TTL signals (`parse_rising_and_falling_frames_from_ttl`). [PR #244](https://github.com/catalystneuro/neuroconv/pull/244)
```
Comment (bendichter) [inline CHANGELOG.md:43]:
```suggestion
* Added a tool for determining rising and falling frames from TTL signals (`parse_rising_frames_from_ttl` and `get_falling_frames_from_ttl`). [PR #244](https://github.com/catalystneuro/neuroconv/pull/244)
```

File: src/neuroconv/tools/signal_processing.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        The frame indices of rising events.
+    """
+    bins = bins or _get_default_bins(trace=trace)
+
+    binned_states = np.digitize(x=trace, bins=bins).astype("int8")
+    diff_binned_states = np.diff(binned_states, axis=0)
+
+    rising_frames = np.where(diff_binned_states > 0)[0]
+
+    return rising_frames
```
Comment (bendichter) [inline src/neuroconv/tools/signal_processing.py:41]:
```suggestion
def get_rising_frames_from_ttl(trace: np.ndarray, threshold: float = None) -> np.ndarray:
    """
    Return the frame indices for rising events in a TTL pulse.
    
    Parameters
    ----------
    trace : numpy.ndarray
        A TTL pulse.
    
    threshold : float, optional
        The mean of trace is used by default.
    
    Returns
    -------
    numpy.ndarray
        The frame indices of rising events.
    """
    
    if threshold is None:
        threshold = np.mean(trace)
        
    return np.where(np.diff(np.sign(trace - threshold)) > 0)[0] + 1
```

@CodyCBakerPhD is there anything wrong this this simpler version of the function?

File: src/neuroconv/tools/signal_processing.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+
+def get_falling_frames_from_ttl(trace: np.ndarray, bins: Optional[np.ndarray] = None) -> Tuple[np.ndarray, np.ndarray]:
```
Comment (bendichter) [inline src/neuroconv/tools/signal_processing.py:44]:
```suggestion
def get_falling_frames_from_ttl(trace: np.ndarray, bins: Optional[np.ndarray] = None) -> np.ndarray:
```

File: src/neuroconv/tools/signal_processing.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    Returns
+    -------
+    rising_frames: numpy.ndarray
+        The frame indices of rising events.
+    """
+    threshold = threshold or np.mean(trace)
+
+    sign = np.sign(trace - threshold)
+    diff = np.diff(sign)
+    rising_frames = np.where(diff > 0)[0]
```
Comment (bendichter) [inline src/neuroconv/tools/signal_processing.py:28]:
```suggestion
    rising_frames = np.where(diff > 0)[0] + 1
```

I think you need to add one here. take a look at a simple trace e.g. `[-1, -1, 1]` to see

File: src/neuroconv/tools/signal_processing.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    Returns
+    -------
+    falling_frames: numpy.ndarray
+        The frame indices of falling events.
+    """
+    threshold = threshold or np.mean(trace)
+
+    sign = np.sign(trace - threshold)
+    diff = np.diff(sign)
+    falling_frames = np.where(diff < 0)[0]
```
Comment (bendichter) [inline src/neuroconv/tools/signal_processing.py:54]:
```suggestion
    falling_frames = np.where(diff < 0)[0] + 1
```

**Review by bendichter — CHANGES_REQUESTED** [review]
see comment

File: src/neuroconv/tools/signal_processing.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    threshold: float, optional
+        The threshold used to distinguish on/off states in the trace.
+        The mean of the trace is used by default.
+
+    Returns
+    -------
+    falling_frames: numpy.ndarray
+        The frame indices of falling events.
+    """
+    threshold = threshold or np.mean(trace)
```
Comment (bendichter) [inline src/neuroconv/tools/signal_processing.py:50]:
```suggestion
    if threshold is None:
         threshold = np.mean(trace)
```

File: src/neuroconv/tools/signal_processing.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    """
+    Return the frame indices for falling events in a TTL pulse.
+
+    Parameters
+    ----------
+    trace: numpy.ndarray
+        A TTL pulse.
+    threshold: float, optional
+        The threshold used to distinguish on/off states in the trace.
+        The mean of the trace is used by default.
```
Comment (bendichter) [inline src/neuroconv/tools/signal_processing.py:43]:
```suggestion
    trace : numpy.ndarray
        A TTL pulse.
    threshold : float, optional
        The threshold used to distinguish on/off states in the trace.
        The mean of the trace is used by default.
```

File: src/neuroconv/tools/signal_processing.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+def get_rising_frames_from_ttl(trace: np.ndarray, threshold: Optional[float] = None) -> np.ndarray:
+    """
+    Return the frame indices for rising events in a TTL pulse.
+
+    Parameters
+    ----------
+    trace: numpy.ndarray
+        A TTL pulse.
+    threshold: float, optional
```
Comment (bendichter) [inline src/neuroconv/tools/signal_processing.py:15]:
```suggestion
    trace : numpy.ndarray
        A TTL pulse.
    threshold : float, optional
```

File: src/neuroconv/tools/signal_processing.py [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+
+def get_rising_frames_from_ttl(trace: np.ndarray, threshold: Optional[float] = None) -> np.ndarray:
+    """
+    Return the frame indices for rising events in a TTL pulse.
+
+    Parameters
+    ----------
+    trace: numpy.ndarray
+        A TTL pulse.
```
Comment (bendichter) [inline src/neuroconv/tools/signal_processing.py:14]:
```suggestion
        A TTL signal.
```

File: src/neuroconv/tools/signal_processing.py [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+
+def get_falling_frames_from_ttl(trace: np.ndarray, threshold: Optional[float] = None) -> np.ndarray:
+    """
+    Return the frame indices for falling events in a TTL pulse.
+
+    Parameters
+    ----------
+    trace: numpy.ndarray
+        A TTL pulse.
```
Comment (bendichter) [inline src/neuroconv/tools/signal_processing.py:40]:
```suggestion
        A TTL signal.
```

File: src/neuroconv/tools/signal_processing.py [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    threshold: float, optional
+        The threshold used to distinguish on/off states in the trace.
+        The mean of the trace is used by default.
+
+    Returns
+    -------
+    rising_frames: numpy.ndarray
```
Comment (bendichter) [inline src/neuroconv/tools/signal_processing.py:21]:
```suggestion
    rising_frames : numpy.ndarray
```

File: src/neuroconv/tools/signal_processing.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    ----------
+    trace: numpy.ndarray
+        A TTL pulse.
+    threshold: float, optional
+        The threshold used to distinguish on/off states in the trace.
+        The mean of the trace is used by default.
+
+    Returns
+    -------
+    falling_frames: numpy.ndarray
```
Comment (bendichter) [inline src/neuroconv/tools/signal_processing.py:47]:
```suggestion
    falling_frames : numpy.ndarray
```

**Review by bendichter — APPROVED** [review]

**Pull request merged**